code/parachain/frame/composable-maths/src/dex/mod.rs

```diff
@@ -1,4 +1,33 @@
+use sp_runtime::{traits::CheckedAdd, PerThing};
+
 pub mod constant_product;
 pub mod price;
 #[cfg(test)]
 pub mod tests;
+
+pub trait WeightMath: PerThing {
+	/// Sums all weights in an array. If weights total greater than `Perthing::one()`, returns
+	/// `None`.
+	fn sum_weights(weights: &[Self]) -> Option<Self>;
+
+	/// Returns `true` if all weights in array are non-zero.
+	fn non_zero_weights(weights: &[Self]) -> bool;
```

Comment:
these fns could take `impl IntoIterator<Item = &Self>` to avoid unnecessary `.collect::<Vec>()` (and avoiding `.copied()`/`.cloned()` by not needing to take ownership of `Self`)